api/prometheus.go

```diff
@@ -0,0 +1,453 @@
+		return BuildMetadataSeriesSet(series)
+	}
+
+	minFrom = util.Min(minFrom, q.from)
+	maxTo = util.Max(maxTo, q.to)
+	for _, s := range series {
+		for _, metric := range s.Series {
+			for _, archive := range metric.Defs {
+				consReq := consolidation.None
+				fn := mdata.Aggregations.Get(archive.AggId).AggregationMethod[0]
+				cons := consolidation.Consolidator(fn)
+
+				newReq := models.NewReq(archive.Id, archive.NameWithTags(), target, q.from, q.to, math.MaxUint32, uint32(archive.Interval), cons, consReq, s.Node, archive.SchemaId, archive.AggId)
+				reqs = append(reqs, newReq)
+			}
+		}
+	}
+
+	select {
+	case <-q.ctx.Done():
+		//request canceled
+		return nil, fmt.Errorf("request canceled")
+	default:
+	}
+
+	reqRenderSeriesCount.Value(len(reqs))
+	if len(reqs) == 0 {
+		return nil, fmt.Errorf("no series found")
+	}
+
+	// note: if 1 series has a movingAvg that requires a long time range extension, it may push other reqs into another archive. can be optimized later
+	reqs, _, _, err = alignRequests(uint32(time.Now().Unix()), minFrom, maxTo, reqs)
+	if err != nil {
+		log.Error(3, "HTTP Render alignReq error: %s", err)
+		return nil, err
+	}
+
+	out, err := q.getTargets(q.ctx, reqs)
+	if err != nil {
+		log.Error(3, "HTTP Render %s", err.Error())
+		return nil, err
+	}
+
+	return SeriesToSeriesSet(out)
+}
+
+// LabelValues returns all potential values for a label name.
+func (q *querier) LabelValues(name string) ([]string, error) {
+	expressions := []string{"name=~[a-zA-Z_][a-zA-Z0-9_]*$"}
+	if name == model.MetricNameLabel {
+		name = "name"
+		expressions = append(expressions, "name=~[a-zA-Z_:][a-zA-Z0-9_:]*$")
+	}
+	result, err := q.MetricIndex.FindTagValues(q.OrgID, name, "", expressions, 0, 100000)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch label values for key:%s, error: %v", name, err)
+	}
+	if result == nil {
+		result = []string{}
+	}
+	return result, nil
+}
+
+// Close releases the resources of the Querier.
+func (q *querier) Close() error {
+	return nil
+}
+
+func SeriesToSeriesSet(out []models.Series) (*models.PrometheusSeriesSet, error) {
+	series := []storage.Series{}
+	for _, metric := range out {
+		series = append(series, models.NewPrometheusSeries(buildTagSet(metric.Target), dataPointsToPrometheusSamplePairs(metric.Datapoints)))
+	}
+	return models.NewPrometheusSeriesSet(series), nil
+}
+
+func dataPointsToPrometheusSamplePairs(data []schema.Point) []model.SamplePair {
+	samples := []model.SamplePair{}
+	for _, point := range data {
+		if math.IsNaN(point.Val) {
+			continue
+		}
+		samples = append(samples, model.SamplePair{
+			Timestamp: model.Time(int64(point.Ts) * 1000),
+			Value:     model.SampleValue(point.Val),
+		})
+	}
+	return samples
+}
+
+func BuildMetadataSeriesSet(seriesNames []Series) (*models.PrometheusSeriesSet, error) {
+	series := []storage.Series{}
+	for _, s := range seriesNames {
+		for _, metric := range s.Series {
+			for _, archive := range metric.Defs {
```

Comment:
note that in MT, the reason why 1 series would have multiple defs, is typically to support the same metric being reported with a different interval. they would have identical tags. so i'm not sure if it's useful listing them all